### PR TITLE
Replacing stateful helpers with stateless helpers

### PR DESCRIPTION
**Pull request by brdandu**
- Replacing chip_client_clusters, chip_cluster_commands, chip_cluster_responses and chip_cluster_command_arguments with all_user_clusters, zcl_commands, zcl_command_responses and zcl_command_arguments respectively in templates
- Updating the required zap version in sdk.json
- Github: ZAP#971

### CODE REVIEW

**Review by andy31415 — APPROVED** [review]

File: src/darwin/Framework/CHIP/templates/MTRClusters-src.zapt [hunk 1]
```diff
@@ -75,7 +76,7 @@ MTR{{cluster}}Cluster{{command}}Params
 {{/inline}}
 {{#*inline "clusterId"}}
 {{#if (wasRemoved cluster command=command)}}
-{{asMEI ../manufacturerCode ../code}}
+{{asMEI ../../manufacturerCode ../../code}}
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/templates/MTRClusters-src.zapt:79]: Why was the change needed?  Where is the extra scope coming from compared to before?

File: src/darwin/Framework/CHIP/templates/MTRClusters.zapt [hunk 2]
```diff
@@ -23,19 +23,21 @@ NS_ASSUME_NONNULL_BEGIN
                               endpointID:(NSNumber *)endpointID
                                    queue:(dispatch_queue_t)queue NS_DESIGNATED_INITIALIZER {{availability (asUpperCamelCase name preserveAcronyms=true) minimalRelease="First major API revamp"}};
 
-{{#chip_cluster_commands}}
+{{#zcl_commands}}
+  {{#if (is_str_equal source 'client')}}
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/templates/MTRClusters.zapt:27]:
```suggestion
{{#if (is_str_equal source 'client')}}
```
to keep indentation from going backwards...

File: src/darwin/Framework/CHIP/templates/MTRClusters.zapt [hunk 3]
```diff
@@ -23,19 +23,21 @@ NS_ASSUME_NONNULL_BEGIN
...
+- (void){{asLowerCamelCase name}}WithParams:(MTR{{cluster}}Cluster{{command}}Params * {{#unless commandHasRequiredField}}_Nullable{{/unless}})params expectedValues:(NSArray<NSDictionary<NSString *, id> *> * _Nullable)expectedDataValueDictionaries expectedValueInterval:(NSNumber * _Nullable)expectedValueIntervalMs completion:({{>command_completion_type command=.}})completion {{availability cluster command=command minimalRelease="First major API revamp"}};
+{{#unless hasArguments}}
 - (void){{asLowerCamelCase name}}WithExpectedValues:(NSArray<NSDictionary<NSString *, id> *> *)expectedValues expectedValueInterval:(NSNumber *)expectedValueIntervalMs completion:({{>command_completion_type command=.}})completion {{availability cluster command=command minimalRelease="First major API revamp"}};
 {{/unless}}
 {{/unless}}
 {{/inline}}
 {{> commandDecl cluster=(asUpperCamelCase parent.name preserveAcronyms=true)
                 command=(asUpperCamelCase name preserveAcronyms=true)}}
-{{/chip_cluster_commands}}
+  {{/if}}
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/templates/MTRClusters.zapt:39]:
```suggestion
{{/if}}
```

File: src/darwin/Framework/CHIP/templates/MTRClusters.zapt [hunk 4]
```diff
@@ -79,20 +81,22 @@ NS_ASSUME_NONNULL_BEGIN
                                endpoint:(uint16_t)endpoint
                                   queue:(dispatch_queue_t)queue {{availability (compatClusterNameRemapping name) deprecatedRelease="First major API revamp" deprecationMessage="Please use initWithDevice:endpoindID:queue:"}};
 
-{{#chip_cluster_commands}}
+{{#zcl_commands}}
+  {{#if (is_str_equal source 'client')}}
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/templates/MTRClusters.zapt:85]:
```suggestion
{{#if (is_str_equal source 'client')}}
```

File: src/darwin/Framework/CHIP/templates/MTRClusters.zapt [hunk 5]
```diff
@@ -79,20 +81,22 @@ NS_ASSUME_NONNULL_BEGIN
...
+{{#unless hasArguments}}
 - (void){{asLowerCamelCase command}}WithExpectedValues:(NSArray<NSDictionary<NSString *, id> *> *)expectedValues expectedValueInterval:(NSNumber *)expectedValueIntervalMs completionHandler:({{>command_completion_type command=. compatRemapNames=true}})completionHandler {{availability cluster command=command deprecatedRelease="First major API revamp" deprecationMessage=(concat "Please use " (asLowerCamelCase name) "WithExpectedValues:expectedValueInterval:completion:")}};
 {{/unless}}
 {{/if}}
 {{/inline}}
 {{> commandDecl cluster=(compatClusterNameRemapping parent.name)
                 command=(compatCommandNameRemapping parent.name name)}}
-{{/chip_cluster_commands}}
+  {{/if}}
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/templates/MTRClusters.zapt:98]:
```suggestion
{{/if}}
```

File: src/darwin/Framework/CHIP/templates/MTRClusters-src.zapt (hunk 1 above)
Comment (brdandu) [reply src/darwin/Framework/CHIP/templates/MTRClusters-src.zapt:79]: It seems to be coming from the arguments to `commandImpl` when used within `zcl_commands`
In `commandImpl` I see the following:
```this : MfgSpecificPing 0x1002 0x0; -> command
parent : MfgSpecificPing 0x1002 0x0; -> command
parent.parent : Basic Information 0x0 0x28; ->cluster```

Comment (bzbarsky-apple) [reply src/darwin/Framework/CHIP/templates/MTRClusters-src.zapt:79]: Yes, but why is it different from the other method of iterating over commands?

Comment (brdandu) [reply src/darwin/Framework/CHIP/templates/MTRClusters-src.zapt]: Filed https://github.com/handlebars-lang/handlebars.js/issues/1931 based on our discussion in https://csamembers.slack.com/archives/C01B01GH64E/p1680186533930189

**Review by bzbarsky-apple — APPROVED** [review]

**Pull request merged**